### ISSUE TITLE
Freenas-api-iscsi driver formatting disk every time instead of just initially

### ISSUE THREAD

**Issue by TripleE-0** [democratic-csi/democratic-csi#215]
Hi! I am newish to Nomad and the Democratic-csi plugin; I had gotten local host folders to work and thought I should switch to network storage since I have TrueNAS Scale. 

Democratic-csi seems to be formatting iscsi disks every time they are mounted in a container, which erases all the data. 

I am using Nomad 1.3.1 with TrueNAS Scale 22.02.2 and the freenas-api-iscsi driver. 

I initially created the container and added a text file to the mounted iscsi folder. I then stopped the container, waited for the node processes to finish, and then started it again. 
When I started it for the second time I noticed that the democratic node had this line in the log file: 
```
executing filesystem command: mkfs.ext4 /dev/sde
```
The mounted iscsi folder did not contain the text file from the first container.  Is there a check for existing file systems that is failing? Otherwise I am not sure why the democratic node would format the iscsi disk every time. 

See the complete logs and configurations below: 

<details>
    <summary>storage_wikijs-db.hcl</summary>

```
id           = "wikijs-db-data"
name         = "wikijs-db-data"
type         = "csi"
plugin_id    = "org.democratic-csi.iscsi"
capacity_max = "5Gib"
capacity_min = "2Gib"

//Suggested to have only one capability
// capability {
//   access_mode     = "single-node-reader-only"
//   attachment_mode = "block-device"
// }

capability {
  access_mode     = "single-node-writer"
  attachment_mode = "file-system"
}

mount_options {
  fs_type     = "ext4"
  mount_flags = ["noatime"]
}
```
</details>

<details>
    <summary>wikijs.job</summary>

```
job "wikijs" {
  datacenters = ["dc1"]
  type        = "service"

  meta {
    version = "23"
  }

  group "Wikijs" {
    update {
      max_parallel = 1
    }

    # Prestart task
    task "set-permissions" {
      driver = "docker"
      volume_mount {
          volume = "wikijs-data"
          destination = "/wiki/data"
          read_only = false
        }

      config {
        image        = "busybox:1.28"
        command      = "sh"
        args         = ["-c", "echo -n 'Setting permissions\n'; chown 1000:1000 -R /wiki/data"]
        network_mode = "bridge"
      }

      resources {
        cpu    = 200
        memory = 128
      }

      lifecycle {
        hook    = "prestart"
        sidecar = false
      }
    }

    # End of Prestart Task


    volume "wikijs-data" {
      type = "csi"
      read_only = false
      source = "wikijs-data"
      attachment_mode = "file-system"
      access_mode     = "single-node-writer"
      mount_options {
        fs_type     = "ext4"
        mount_flags = ["noatime"]
      }
    }

    volume "wikijs-db-data" {
      type = "csi"
      read_only = false
      source = "wikijs-db-data"
      attachment_mode = "file-system"
      access_mode     = "single-node-writer"
      mount_options {
        fs_type     = "ext4"
        mount_flags = ["noatime"]
      }
    }
    
    task "wiki" {
        driver = "docker"
        volume_mount {
          volume = "wikijs-data"
          destination = "/wiki/data"
          read_only = false
        } 
        config {
            image = "ghcr.io/requarks/wiki:2"
            port_map {
                http = 3000
            }

        }

        env {
            DB_TYPE = "postgres"
            DB_HOST = "${NOMAD_IP_http}"
            #DB_HOST = "wiki-db.services.dc1.consul"
            #DB_HOST = "{{ env "NOMAD_IP_http" }}wiki-db.homelab"
            DB_PORT = "25432"
            DB_USER = "postgres"
            DB_PASS = "<password here>"
            DB_NAME = "wikijs"
        }

    service {
        name = "wiki"
        port = "http"
        tags = [
          "traefik.enable=true",
          "traefik.http.middlewares.httpsRedirect.redirectscheme.scheme=https",
          "traefik.http.routers.${NOMAD_TASK_NAME}_insecure.middlewares=httpsRedirect",
          "traefik.http.routers.${NOMAD_TASK_NAME}_insecure.rule=Host(`${NOMAD_TASK_NAME}`)",
          "traefik.http.routers.${NOMAD_TASK_NAME}.tls.domains[0].sans=${NOMAD_TASK_NAME}",
          "traefik.http.routers.${NOMAD_TASK_NAME}.tls.certresolver=myresolver",
          "traefik.http.routers.${NOMAD_TASK_NAME}.middlewares=httpsRedirect"
        ]
        check {
          type     = "http"
          path     = "/"
          #port     = "http"
          interval = "30s"
          timeout  = "2s"
        }
      }

      resources {
        cpu    = 500
        memory = 512
        network {
          port  "http"  {}
        }
      }
    }

    
    task "wiki-db" {
      driver = "docker"

      volume_mount {
          volume = "wikijs-db-data"
          destination = "/appdata"
          read_only = false
        }

      #user = "postgres"
      config {
        image = "postgres:14-bullseye"
        

        port_map {
          db = 5432
        }
      }

      env {
        POSTGRES_DB       = "wikijs"
        POSTGRES_USER     = "postgres"
        POSTGRES_PASSWORD = "<password here>"
        PGPASSWORD        = "<password here>"
        PGDATA            = "/appdata/db/postgres/"
      }

      service {
        name = "${NOMAD_TASK_NAME}"
        tags = ["postgres"]

        port = "db"

        check {
          type     = "tcp"
          port     = "db"
          interval = "30s"
          timeout  = "2s"
        }
      }

      resources {
        network {
          port  "db"  {
            static = "25432"
          }
        }
      }
    }
    


    
  }
}

```
</details>

<details>
    <summary>democratic-csi iscsi node logs</summary>

These are the logs from the node that was hosting this container. This issue affects all of the nodes. 

Initial creation of container
```
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeStageVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"node_attach_driver\":\"iscsi\",\"portal\":\"192.168.1.130:3260\",\"portals\":\"\",\"interface\":\"\",\"iqn\":\"iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera\",\"lun\":\"0\",\"provisioner_driver\":\"freenas-api-iscsi\"},\"volume_id\":\"wikijs-db-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"ext4\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"}},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:40:06.066Z"}
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera -p 192.168.1.130:3260 -o new
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera -p 192.168.1.130:3260 -o update --name node.startup --value manual
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera -p 192.168.1.130:3260 -l
executing iscsi command: iscsiadm -m session
executing iscsi command: iscsiadm -m session -r 336 --rescan
retry - failed condition, not trying again
executing filesystem command: realpath /dev/disk/by-path/ip-192.168.1.130:3260-iscsi-iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera-lun-0
{"host":"nomad4","level":"info","message":"successfully logged into portal 192.168.1.130:3260 and created device /dev/disk/by-path/ip-192.168.1.130:3260-iscsi-iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera-lun-0 with realpath /dev/sdc","service":"democratic-csi","timestamp":"2022-07-01T18:40:08.189Z"}
executing filesystem command: sh -c for file in $(ls -la /dev/mapper/* | grep "\->" | grep -oP "\-> .+" | grep -oP " .+"); do echo $(F=$(echo $file | grep -oP "[a-z0-9-]+");echo $F":"$(ls "/sys/block/${F}/slaves/");); done;
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -J -O /dev/sdc
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -J -O /dev/sdc
executing filesystem command: mkfs.ext4 /dev/sdc
executing filesystem command: blkid -p -o export /dev/sdc
executing filesystem command: realpath /dev/sdc
executing mount command: findmnt --source /dev/sdc --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sdc
executing mount command: findmnt --source /dev/sdc --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing mount command: mount -t ext4 -o noatime,defaults /dev/sdc /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: resize2fs /dev/sdc
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeStageVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:40:10.457Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodePublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"portal\":\"192.168.1.130:3260\",\"portals\":\"\",\"interface\":\"\",\"iqn\":\"iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera\",\"lun\":\"0\",\"provisioner_driver\":\"freenas-api-iscsi\",\"node_attach_driver\":\"iscsi\"},\"volume_id\":\"wikijs-db-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer\",\"target_path\":\"/local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"ext4\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"},\"readonly\":false},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:40:10.458Z"}
retry - failed condition, not trying again
executing filesystem command: mkdir -p -m 0750 /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: mount --bind -o noatime,defaults /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodePublishVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:40:10.515Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeStageVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"provisioner_driver\":\"freenas-api-iscsi\",\"node_attach_driver\":\"iscsi\",\"portal\":\"192.168.1.130:3260\",\"portals\":\"\",\"interface\":\"\",\"iqn\":\"iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera\",\"lun\":\"0\"},\"volume_id\":\"wikijs-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-data/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"ext4\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"}},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:40:10.518Z"}
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera -p 192.168.1.130:3260 -o new
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera -p 192.168.1.130:3260 -o update --name node.startup --value manual
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera -p 192.168.1.130:3260 -l
executing iscsi command: iscsiadm -m session
executing iscsi command: iscsiadm -m session -r 337 --rescan
retry - failed condition, not trying again
executing filesystem command: realpath /dev/disk/by-path/ip-192.168.1.130:3260-iscsi-iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera-lun-0
{"host":"nomad4","level":"info","message":"successfully logged into portal 192.168.1.130:3260 and created device /dev/disk/by-path/ip-192.168.1.130:3260-iscsi-iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera-lun-0 with realpath /dev/sde","service":"democratic-csi","timestamp":"2022-07-01T18:40:12.625Z"}
executing filesystem command: sh -c for file in $(ls -la /dev/mapper/* | grep "\->" | grep -oP "\-> .+" | grep -oP " .+"); do echo $(F=$(echo $file | grep -oP "[a-z0-9-]+");echo $F":"$(ls "/sys/block/${F}/slaves/");); done;
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -J -O /dev/sde
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -J -O /dev/sde
executing filesystem command: mkfs.ext4 /dev/sde
executing filesystem command: blkid -p -o export /dev/sde
executing filesystem command: realpath /dev/sde
executing mount command: findmnt --source /dev/sde --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sde
executing mount command: findmnt --source /dev/sde --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing mount command: mount -t ext4 -o noatime,defaults /dev/sde /local/csi/staging/wikijs-data/rw-file-system-single-node-writer
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: resize2fs /dev/sde
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeStageVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:40:14.873Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodePublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"node_attach_driver\":\"iscsi\",\"portal\":\"192.168.1.130:3260\",\"portals\":\"\",\"interface\":\"\",\"iqn\":\"iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera\",\"lun\":\"0\",\"provisioner_driver\":\"freenas-api-iscsi\"},\"volume_id\":\"wikijs-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-data/rw-file-system-single-node-writer\",\"target_path\":\"/local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-data/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"ext4\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"},\"readonly\":false},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:40:14.874Z"}
retry - failed condition, not trying again
executing filesystem command: mkdir -p -m 0750 /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-data/rw-file-system-single-node-writer
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: mount --bind -o noatime,defaults /local/csi/staging/wikijs-data/rw-file-system-single-node-writer /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-data/rw-file-system-single-node-writer
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodePublishVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:40:14.929Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: Probe call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:40:18.951Z"}
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: Probe response: {\"ready\":{\"value\":true}}","service":"democratic-csi","timestamp":"2022-07-01T18:40:18.951Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: Probe call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:40:33.082Z"}
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: Probe response: {\"ready\":{\"value\":true}}","service":"democratic-csi","timestamp":"2022-07-01T18:40:33.082Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeGetCapabilities call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:40:33.085Z"}
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeGetCapabilities response: {\"capabilities\":[{\"rpc\":{\"type\":\"STAGE_UNSTAGE_VOLUME\"}},{\"rpc\":{\"type\":\"GET_VOLUME_STATS\"}},{\"rpc\":{\"type\":\"EXPAND_VOLUME\"}},{\"rpc\":{\"type\":\"SINGLE_NODE_MULTI_WRITER\"}}]}","service":"democratic-csi","timestamp":"2022-07-01T18:40:33.085Z"}

```

Stopping container 
```
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeUnpublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"volume_id\":\"wikijs-data\",\"target_path\":\"/local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-data/rw-file-system-single-node-writer\"},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.244Z"}
executing mount command: findmnt --mountpoint /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeUnpublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"volume_id\":\"wikijs-db-data\",\"target_path\":\"/local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer\"},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.252Z"}
executing mount command: findmnt --mountpoint /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: umount /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-data/rw-file-system-single-node-writer
executing mount command: umount /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer
executing filesystem command: rmdir /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-data/rw-file-system-single-node-writer
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeUnpublishVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.285Z"}
executing filesystem command: rmdir /local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeUnstageVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"volume_id\":\"wikijs-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-data/rw-file-system-single-node-writer\"},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.289Z"}
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer/block_device --output source,target,fstype,label,options -b -J --nofsroot
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeUnpublishVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.292Z"}
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeUnstageVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"volume_id\":\"wikijs-db-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer\"},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.301Z"}
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer/block_device --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sde
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -J -O /dev/sde
executing filesystem command: realpath /dev/sda
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -J -O /dev/sdc
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: umount /local/csi/staging/wikijs-data/rw-file-system-single-node-writer
executing mount command: umount /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: realpath /dev/sde
executing filesystem command: realpath /dev/sde
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -J -O /dev/sde
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -J -O /dev/sde
executing filesystem command: realpath /dev/sda
executing filesystem command: realpath /dev/sdc
executing iscsi command: iscsiadm -m session -P 3
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -J -O /dev/sdc
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -J -O /dev/sdc
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera -p 192.168.1.130:3260,1 -u
executing iscsi command: iscsiadm -m session -P 3
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera -p 192.168.1.130:3260,1 -u
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera -p 192.168.1.130:3260,1 -o delete
executing filesystem command: rmdir /local/csi/staging/wikijs-data/rw-file-system-single-node-writer
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeUnstageVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.568Z"}
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera -p 192.168.1.130:3260,1 -o delete
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeUnpublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"volume_id\":\"wikijs-db-data\",\"target_path\":\"/local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer\"},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.578Z"}
{"host":"nomad4","level":"error","message":"handler error - driver: FreeNASApiDriver method: NodeUnpublishVolume error: {\"name\":\"GrpcError\",\"code\":10,\"message\":\"operation locked due to in progress operation(s): [\\\"volume_id_wikijs-db-data\\\"]\"}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.578Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeUnpublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"volume_id\":\"wikijs-db-data\",\"target_path\":\"/local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer\"},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.580Z"}
{"host":"nomad4","level":"error","message":"handler error - driver: FreeNASApiDriver method: NodeUnpublishVolume error: {\"name\":\"GrpcError\",\"code\":10,\"message\":\"operation locked due to in progress operation(s): [\\\"volume_id_wikijs-db-data\\\"]\"}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.580Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeUnpublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"volume_id\":\"wikijs-db-data\",\"target_path\":\"/local/csi/per-alloc/979f2355-c806-90e6-0145-e7f9af2111f7/wikijs-db-data/rw-file-system-single-node-writer\"},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.585Z"}
{"host":"nomad4","level":"error","message":"handler error - driver: FreeNASApiDriver method: NodeUnpublishVolume error: {\"name\":\"GrpcError\",\"code\":10,\"message\":\"operation locked due to in progress operation(s): [\\\"volume_id_wikijs-db-data\\\"]\"}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.585Z"}
executing filesystem command: rmdir /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeUnstageVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:42:21.590Z"}

```

Starting container
```
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: Probe response: {\"ready\":{\"value\":true}}","service":"democratic-csi","timestamp":"2022-07-01T18:45:18.972Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeStageVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"provisioner_driver\":\"freenas-api-iscsi\",\"node_attach_driver\":\"iscsi\",\"portal\":\"192.168.1.130:3260\",\"portals\":\"\",\"interface\":\"\",\"iqn\":\"iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera\",\"lun\":\"0\"},\"volume_id\":\"wikijs-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-data/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"ext4\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"}},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:45:23.092Z"}
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera -p 192.168.1.130:3260 -o new
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera -p 192.168.1.130:3260 -o update --name node.startup --value manual
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera -p 192.168.1.130:3260 -l
executing iscsi command: iscsiadm -m session
executing iscsi command: iscsiadm -m session -r 338 --rescan
retry - failed condition, not trying again
executing filesystem command: realpath /dev/disk/by-path/ip-192.168.1.130:3260-iscsi-iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera-lun-0
{"host":"nomad4","level":"info","message":"successfully logged into portal 192.168.1.130:3260 and created device /dev/disk/by-path/ip-192.168.1.130:3260-iscsi-iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera-lun-0 with realpath /dev/sdc","service":"democratic-csi","timestamp":"2022-07-01T18:45:26.272Z"}
executing filesystem command: sh -c for file in $(ls -la /dev/mapper/* | grep "\->" | grep -oP "\-> .+" | grep -oP " .+"); do echo $(F=$(echo $file | grep -oP "[a-z0-9-]+");echo $F":"$(ls "/sys/block/${F}/slaves/");); done;
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -J -O /dev/sdc
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -J -O /dev/sdc
executing filesystem command: mkfs.ext4 /dev/sdc
executing filesystem command: blkid -p -o export /dev/sdc
executing filesystem command: realpath /dev/sdc
executing mount command: findmnt --source /dev/sdc --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sdc
executing mount command: findmnt --source /dev/sdc --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing mount command: mount -t ext4 -o noatime,defaults /dev/sdc /local/csi/staging/wikijs-data/rw-file-system-single-node-writer
executing filesystem command: realpath /dev/sdc
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: resize2fs /dev/sdc
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeStageVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:45:28.553Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodePublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"lun\":\"0\",\"provisioner_driver\":\"freenas-api-iscsi\",\"node_attach_driver\":\"iscsi\",\"portal\":\"192.168.1.130:3260\",\"portals\":\"\",\"interface\":\"\",\"iqn\":\"iqn.2005-10.org.freenas.ctl:csi-wikijs-data-clustera\"},\"volume_id\":\"wikijs-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-data/rw-file-system-single-node-writer\",\"target_path\":\"/local/csi/per-alloc/7df45fbd-0ecf-dc5c-72ae-e843c25feff8/wikijs-data/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"ext4\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"},\"readonly\":false},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:45:28.555Z"}
retry - failed condition, not trying again
executing filesystem command: mkdir -p -m 0750 /local/csi/per-alloc/7df45fbd-0ecf-dc5c-72ae-e843c25feff8/wikijs-data/rw-file-system-single-node-writer
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /local/csi/per-alloc/7df45fbd-0ecf-dc5c-72ae-e843c25feff8/wikijs-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: mount --bind -o noatime,defaults /local/csi/staging/wikijs-data/rw-file-system-single-node-writer /local/csi/per-alloc/7df45fbd-0ecf-dc5c-72ae-e843c25feff8/wikijs-data/rw-file-system-single-node-writer
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodePublishVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:45:28.610Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeStageVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"interface\":\"\",\"iqn\":\"iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera\",\"lun\":\"0\",\"provisioner_driver\":\"freenas-api-iscsi\",\"node_attach_driver\":\"iscsi\",\"portal\":\"192.168.1.130:3260\",\"portals\":\"\"},\"volume_id\":\"wikijs-db-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"ext4\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"}},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:45:28.612Z"}
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera -p 192.168.1.130:3260 -o new
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera -p 192.168.1.130:3260 -o update --name node.startup --value manual
executing iscsi command: iscsiadm -m node -T iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera -p 192.168.1.130:3260 -l
executing iscsi command: iscsiadm -m session
executing iscsi command: iscsiadm -m session -r 339 --rescan
retry - failed condition, not trying again
executing filesystem command: realpath /dev/disk/by-path/ip-192.168.1.130:3260-iscsi-iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera-lun-0
{"host":"nomad4","level":"info","message":"successfully logged into portal 192.168.1.130:3260 and created device /dev/disk/by-path/ip-192.168.1.130:3260-iscsi-iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera-lun-0 with realpath /dev/sde","service":"democratic-csi","timestamp":"2022-07-01T18:45:30.709Z"}
executing filesystem command: sh -c for file in $(ls -la /dev/mapper/* | grep "\->" | grep -oP "\-> .+" | grep -oP " .+"); do echo $(F=$(echo $file | grep -oP "[a-z0-9-]+");echo $F":"$(ls "/sys/block/${F}/slaves/");); done;
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -J -O /dev/sde
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -J -O /dev/sde
executing filesystem command: mkfs.ext4 /dev/sde
executing filesystem command: blkid -p -o export /dev/sde
executing filesystem command: realpath /dev/sde
executing mount command: findmnt --source /dev/sde --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sde
executing mount command: findmnt --source /dev/sde --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing mount command: mount -t ext4 -o noatime,defaults /dev/sde /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer
executing filesystem command: realpath /dev/sde
executing filesystem command: lsblk -a -b -l -J -O
executing filesystem command: realpath /dev/sda
executing filesystem command: resize2fs /dev/sde
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodeStageVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:45:33.004Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodePublishVolume call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{\"publish_context\":{},\"secrets\":\"redacted\",\"volume_context\":{\"lun\":\"0\",\"provisioner_driver\":\"freenas-api-iscsi\",\"node_attach_driver\":\"iscsi\",\"portal\":\"192.168.1.130:3260\",\"portals\":\"\",\"interface\":\"\",\"iqn\":\"iqn.2005-10.org.freenas.ctl:csi-wikijs-db-data-clustera\"},\"volume_id\":\"wikijs-db-data\",\"staging_target_path\":\"/local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer\",\"target_path\":\"/local/csi/per-alloc/7df45fbd-0ecf-dc5c-72ae-e843c25feff8/wikijs-db-data/rw-file-system-single-node-writer\",\"volume_capability\":{\"access_mode\":{\"mode\":\"SINGLE_NODE_WRITER\"},\"mount\":{\"mount_flags\":[\"noatime\"],\"fs_type\":\"ext4\",\"volume_mount_group\":\"\"},\"access_type\":\"mount\"},\"readonly\":false},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:45:33.007Z"}
retry - failed condition, not trying again
executing filesystem command: mkdir -p -m 0750 /local/csi/per-alloc/7df45fbd-0ecf-dc5c-72ae-e843c25feff8/wikijs-db-data/rw-file-system-single-node-writer
executing mount command: findmnt --mountpoint /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: findmnt --mountpoint /local/csi/per-alloc/7df45fbd-0ecf-dc5c-72ae-e843c25feff8/wikijs-db-data/rw-file-system-single-node-writer --output source,target,fstype,label,options -b -J --nofsroot
executing mount command: mount --bind -o noatime,defaults /local/csi/staging/wikijs-db-data/rw-file-system-single-node-writer /local/csi/per-alloc/7df45fbd-0ecf-dc5c-72ae-e843c25feff8/wikijs-db-data/rw-file-system-single-node-writer
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: NodePublishVolume response: {}","service":"democratic-csi","timestamp":"2022-07-01T18:45:33.071Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: Probe call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:45:33.110Z"}
{"host":"nomad4","level":"info","message":"new response - driver: FreeNASApiDriver method: Probe response: {\"ready\":{\"value\":true}}","service":"democratic-csi","timestamp":"2022-07-01T18:45:33.110Z"}
{"host":"nomad4","level":"info","message":"new request - driver: FreeNASApiDriver method: NodeGetCapabilities call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.45.0\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2022-07-01T18:45:33.111Z"}
```

</details>

[freenas-api-iscsi.yaml.txt](https://github.com/democratic-csi/democratic-csi/files/9065425/freenas-api-iscsi.yaml.txt)

**Comment by travisghansen** [comment]
What version of the driver are you running?

**Comment by travisghansen** [comment]
OK, I can't reproduce locally (thankfully!, this would be a very nasty bug). So the logic behind that check is:

Get the output from the following `lsblk` command (obviously replace the device with the correct/relevant device)

```
lsblk -a -b -J -O /dev/sda
{
   "blockdevices": [
      {
         "alignment": 0,
         "disc-aln": 0,
         "dax": false,
         "disc-gran": 0,
         "disc-max": 0,
         "disc-zero": false,
         "fsavail": "950214656",
         "fsroots": [
             "/"
         ],
         "fssize": "1020702720",
         "fstype": "ext4",
         "fsused": "24576",
         "fsuse%": "0%",
         "fsver": "1.0",
         "group": "disk",
         "hctl": "0:0:0:0",
         "hotplug": false,
         "kname": "sda",
         "label": null,
         "log-sec": 512,
         "maj:min": "8:0",
         "min-io": 16384,
         "mode": "brw-rw----",
         "model": "pvc-111",
         "name": "sda",
         "opt-io": 16384,
         "owner": "root",
         "partflags": null,
         "partlabel": null,
         "parttype": null,
         "parttypename": null,
         "partuuid": null,
         "path": "/dev/sda",
         "phy-sec": 16384,
         "pkname": null,
         "pttype": null,
         "ptuuid": null,
         "ra": 128,
         "rand": false,
         "rev": "4.0 ",
         "rm": false,
         "ro": false,
         "rota": false,
         "rq-size": 226,
         "sched": "mq-deadline",
         "serial": "50e6859b-e255-4609-a366-c7f4e4ce88f3",
         "size": 1073741824,
         "start": null,
         "state": "running",
         "subsystems": "block:scsi",
         "mountpoint": "/mnt/staged",
         "mountpoints": [
             "/mnt/staged"
         ],
         "tran": "iscsi",
         "type": "disk",
         "uuid": "9693fe57-11ed-4484-93fc-7d3763628136",
         "vendor": "LIO-ORG ",
         "wsame": 0,
         "wwn": "0x600140550e6859be2554609a366c7f4e",
         "zoned": "none",
         "zone-sz": 0,
         "zone-wgran": 0,
         "zone-app": 0,
         "zone-nr": 0,
         "zone-omax": 0,
         "zone-amax": 0
      }
   ]
}

Then check the existence of the `fstype` property of the **first** result in the array.

Can you run the above command from inside the container after mounting a device and send over the output?

**Comment by TripleE-0** [comment]
So just to be clear, `/dev` needs to be mounted in the *application* container? The container seems to have `/dev` mounted already as a tmpfs : `tmpfs on /dev type tmpfs (rw,nosuid,seclabel,size=65536k,mode=755)`
In the democratic node I have the root folder `/` mounted as `/host` and that can run the `lsblk -a -b -J -O /dev/sdd` command successfully.

I looked for selinux alerts since the host is Rocky linux 8, but I do not see anything in the logs even after restarting the container.

**Comment by travisghansen** [comment]
Can you send over the node/controller hcl files?

You need to run the container as privileged, host networking, etc. In terms of mounts this is what the k8s charts does: https://github.com/democratic-csi/charts/blob/master/stable/democratic-csi/templates/node.yaml#L147-L181

I don't think these are actually necessary:

```
        - name: iscsi-dir
          mountPath: /etc/iscsi
          mountPropagation: Bidirectional
        - name: iscsi-info
          mountPath: /var/lib/iscsi
          mountPropagation: Bidirectional
```

So probably `/dev`, `/sys` and `/run/udev` (along with `/host` and everything else you already have) should do the trick.

**Comment by TripleE-0** [comment]
Here is the controller I only mount the csi socket. 
<details>
    <summary>storage-controller.job</summary>

```
job "storage-controller" {
  #region = "home"
  datacenters = ["dc1"]
  type = "service"

  #type = "system"

  constraint {
    operator = "distinct_hosts"
    value = true
  }
  
  group "controller" {
    /*
    network {
            mode = "bridge"
            port "grpc" {
              static = 9000
              to     = 9000
            }
        }
        */
    task "controller" {
      driver = "docker"

      config {
        image = "democraticcsi/democratic-csi:latest"
        #ports = ["grpc",]

        args = [
          "--csi-version=1.5.0",
          "--csi-name=org.democratic-csi.iscsi",
          "--driver-config-file=${NOMAD_TASK_DIR}/driver-config-file.yaml",
          "--log-level=verbose",
          "--csi-mode=controller",
          "--server-socket=/csi/csi.sock",
          #"--server-address=0.0.0.0",
          #"--server-port=9000",
        ]

        privileged = true
      }

      env {
        CSI_ENDPOINT = "/csi/csi.sock"
      }

      resources {
        cpu    = 500
        memory = 256
        
      }
    
      csi_plugin {
        id        = "org.democratic-csi.iscsi"
        type      = "controller"
        mount_dir = "/csi"
      }

      template {
        destination = "${NOMAD_TASK_DIR}/driver-config-file.yaml"

        data = <<EOH
< freenas-api-iscsi.yaml goes here >
EOH
      }

      

    }
  }
}

```

</details>

For the democratic node container I set the container to privileged and mount the host file system in `/host` . 

<details>
    <summary>storage-node.job</summary>

```
job "democratic-csi-iscsi-node" {
  datacenters = ["dc1"]

  # you can run node plugins as service jobs as well, but this ensures
  # that all nodes in the DC have a copy.
  type = "system"

  group "nodes" {
    task "plugin" {
      driver = "docker"

      env {
        CSI_NODE_ID = "${attr.unique.hostname}"
      }

      config {
        image = "docker.io/democraticcsi/democratic-csi:latest"

        args = [
          "--csi-version=1.5.0",
          # must match the csi_plugin.id attribute below
          "--csi-name=org.democratic-csi.iscsi",
          "--driver-config-file=${NOMAD_TASK_DIR}/driver-config-file.yaml",
          "--log-level=info",
          "--csi-mode=node",
          "--server-socket=/csi/csi.sock",
        ]

        # node plugins must run as privileged jobs because they
        # mount disks to the host
        privileged = true
        ipc_mode = "host"
        network_mode = "host"

        mount {
          type = "bind"
          target = "/host"
          source = "/"
          readonly=false
        }
      }

      template {
        destination = "${NOMAD_TASK_DIR}/driver-config-file.yaml"

        data = <<EOH
< freenas-api-iscsi.yaml goes here >
EOH
      }

      csi_plugin {
        # must match --csi-name arg
        id        = "org.democratic-csi.iscsi"
        type      = "node"
        mount_dir = "/csi"
      }

      resources {
        cpu    = 500
        memory = 256
      }
    }
  }
}
```

</details>

Seems like I am missing quite a few mounts on the node side.
I will add the mounts to the node job file and back with you. As a side note, thank you for your help and quick responses!

**Comment by TripleE-0** [comment]
After taking a look at that template file, it appears as if line 174 and 175 where flipped by accident? Unless the order doesn't matter. However all the other mounts have the name then the mount path. 
https://github.com/democratic-csi/charts/blob/master/stable/democratic-csi/templates/node.yaml#L174-L175

---

After some testing I was unable to get the `/dev` folder mounted in the democratic node due to error 
`failed to create container: API error (400): Duplicate mount point: /dev`.  Once I commented out the `/dev` mount the node launched with no issue though! 
The application container also has the same data after it moved hosts! 

Is it a problem to not have the `/dev` directory mounted in the node container?

**Comment by travisghansen** [comment]
If you don’t have it mounted then you have the original issue no?

**Comment by TripleE-0** [comment]
The only thing that I changed in the node configuration was adding the `/sys` and `/run/udev` mounts in the container. 

I tested everything by restarting the container and watched it move to another host. Once on the new host it started up with all of the customizations that I had previously made.

Here is the config: 

<details>
    <summary>democratic-node.job</summary>

```
job "democratic-csi-iscsi-node" {
  datacenters = ["dc1"]

  # you can run node plugins as service jobs as well, but this ensures
  # that all nodes in the DC have a copy.
  type = "system"

  group "nodes" {
    update {
      auto_revert      = true
    }
    task "plugin" {
      driver = "docker"
     
      env {
        CSI_NODE_ID = "${attr.unique.hostname}"
      }

      config {
        image = "docker.io/democraticcsi/democratic-csi:latest"
        volumes = [
          #"/dev:/dev",
          "/sys:/sys",
          "/run/udev:/run/udev"
        ]
        args = [
          "--csi-version=1.5.0",
          # must match the csi_plugin.id attribute below
          "--csi-name=org.democratic-csi.iscsi",
          "--driver-config-file=${NOMAD_TASK_DIR}/driver-config-file.yaml",
          "--log-level=info",
          "--csi-mode=node",
          "--server-socket=/csi/csi.sock",
        ]

        # node plugins must run as privileged jobs because they
        # mount disks to the host
        privileged = true
        ipc_mode = "host"
        network_mode = "host"

        // mount {
        //   type = "bind"
        //   target = "/dev"
        //   source = "/dev"
        //   readonly=false
        // }
        // mount {
        //   type = "bind"
        //   target = "/sys"
        //   source = "/sys"
        //   readonly=false
        // }
        // mount {
        //   type = "bind"
        //   target = "/run/udev"
        //   source = "/run/udev"
        //   readonly=false
        // }
        mount {
          type = "bind"
          target = "/host"
          source = "/"
          readonly=false
        }


      }

      template {
        destination = "${NOMAD_TASK_DIR}/driver-config-file.yaml"

        data = <<EOH
< freenas-api-iscsi.yaml here >
EOH
      }

      csi_plugin {
        # must match --csi-name arg
        id        = "org.democratic-csi.iscsi"
        type      = "node"
        mount_dir = "/csi"
      }

      resources {
        cpu    = 500
        memory = 256
      }
    }
  }
}

```

</details>

**Comment by travisghansen** [comment]
Yeah really strange. I already have updated code to prevent the behavior you saw should it ever happen in the future however because that is very egregious.

**Comment by travisghansen** [comment]
At this point let’s just leave this open for a bit longer while you do some more testing and see how it goes. I’ll close it after I commit the fix and it goes out in a tagged release.

**Comment by TripleE-0** [comment]
Sounds good Travis I really appreciate your help.

**Comment by Cellivar** [comment]
Just wanted to chime in that we're observing this same behavior on our system.

Nomad version 1.3.2
Democratic-CSI 1.7.2 (upgraded today from 1.4.2)
Driver configuration: `node-manual`

We've rolled back to 1.4.2 to stop the data loss incidents before we had a chance to do any debugging.

**Comment by travisghansen** [comment]
Wait it worked in 1.4.2 without the bind mount?

**Comment by binaryf0x** [comment]
> Wait it worked in 1.4.2 without the bind mount?

We had other problems (mounts left behind after the iSCSI volume was detached) so it might just be that this only made things sufficiently broken that we discovered the root cause.

**Comment by travisghansen** [comment]
Are you certain the udev mount fixes the issue? Or could it be something else?

**Comment by Cellivar** [comment]
> > Wait it worked in 1.4.2 without the bind mount?
> 
> We had other problems (mounts left behind after the iSCSI volume was detached) so it might just be that this only made things sufficiently broken that we discovered the root cause.

We're still uncertain if this issue was caused specifically by Nomad's handling of CSI in version 1.2.0 or if it was democratic losing the plot. Either way this is what our node job has looked like for nearly two years.

<details>
<summary>csi-democratic-node.nomad.tmpl</summary>

```
job "csi-democrat-node" {
  datacenters = ["dc1"]
  type        = "system"

  group "nodes" {

    restart {
      attempts = 10
      delay = "30s"
      interval = "6m"
      mode = "delay"
    }

    task "plugin" {

      driver = "docker"

      config {
        image = "democraticcsi/democratic-csi:v1.4.2"

        args = [
          "--csi-version=1.2.0",
          "--csi-name=org.democratic-csi.nfs",
          "--driver-config-file=$${NOMAD_TASK_DIR}/driver-config-file.yaml",
          "--log-level=debug",
          "--csi-mode=node",
          "--server-socket=/csi-data/csi.sock",
        ]

        # Nodes use this to set up iscsi volumes
        mount {
          type = "bind"
          target = "/host"
          source = "/"
          readonly=false
        }
        # Nodes use this to communicate with iscsi configs
        mount {
          type = "bind"
          target = "/etc/iscsi"
          source = "/etc/iscsi"
          readonly = false
        }

        # Nodes use this to control iscsiadm
        network_mode = "host"
        ipc_mode = "host"
        privileged = true
      }

      csi_plugin {
        id        = "democratnfs"
        type      = "node"
        mount_dir = "/csi-data"
      }

      template {
        destination = "$${NOMAD_TASK_DIR}/driver-config-file.yaml"

        data = <<EOH
driver: node-manual
EOH
      }

      resources {
        cpu    = 30
        memory = 100
      }
    }
  }
}
```
</details>

Today we:

1. Upgraded Nomad to 1.3.2, which immediately resolved some CSI issues we've had in the past.
2. Confirmed everything was working as expected, including moving volume mounts between machines.
3. Upgraded democratic to 1.7.2 for controllers and nodes.
4. Observed one rescheduled task flapping because of an entirely empty volume mount.
5. Manually forced another task to move to another machine, causing the same issue with a now empty mount and finding the reformat command in the logs for the democratic node job.
6. Found this issue and rolled back to 1.4.2
8. Ran some of the debug commands in this issue, observing as well that lsblk was empty.
9. Added the mount for /run/udev in preparation for the release you mentioned that should at least prevent data loss.

We're going to wait on 1.4.2 as it appears to be working fine for us for the moment, and we don't make use of any of the newer features added in later releases anyhow.

**Comment by travisghansen** [comment]
Fair enough. Super strange to me but I’ll publish a new version asap with the ‘fix’ and then if you could do further testing that would be great.

**Comment by travisghansen** [comment]
Couple things:

- I'll have a new image tagged `next` in a few minutes that I would love for you to try before I actually snap the release
- doing an `strace` on lsblk (in older versions and newer versions reads libudev and also opens files in `/sys`

Can you try running these commands directly on the nodes? For each one `ls /dev`, `ls /sys/block`, and `lsblk` and sanity check them for the relevant files/output/etc:

```
docker run -ti --privileged --entrypoint bash democraticcsi/democratic-csi:v1.4.2
docker run -ti --privileged --entrypoint bash democraticcsi/democratic-csi:v1.7.2
```

I'm really quite baffled why this has come up. I'm wondering if Nomad has changed something with its defaults?

**Comment by travisghansen** [comment]
`next` images are built with the fix.

**Comment by binaryf0x** [comment]
Using `strace` I've confirmed that in both the v1.4.2 and v1.7.2 images `lsblk -o FSTYPE /dev/sdf` will attempt to read from `/run/udev` in order to read the fstype property. If this path isn't mounted from the host then it fails repeatedly and the property comes back as null:

```
readlinkat(AT_FDCWD, "/sys/devices/platform/host5/session13/target5:0:0/5:0:0:1/block/sdf/subsystem", "../../../../../../../../class/bl"..., 4096) = 35
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/run/udev/data/b8:80", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
close(3)                                = 0
ioctl(1, TCGETS, {B38400 opost isig icanon echo ...}) = 0
openat(AT_FDCWD, "/usr/lib/aarch64-linux-gnu/gconv/gconv-modules.cache", O_RDONLY) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=26404, ...}) = 0
mmap(NULL, 26404, PROT_READ, MAP_SHARED, 3, 0) = 0xffff9be1e000
close(3)                                = 0
futex(0xffff9c391830, FUTEX_WAKE_PRIVATE, 2147483647) = 0
fstat(1, {st_mode=S_IFCHR|0620, st_rdev=makedev(0x88, 0), ...}) = 0
write(1, "FSTYPE\n", 7FSTYPE
)                 = 7
write(1, "\n", 1
)                       = 1
close(1)                                = 0
close(2)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

The code from `lsblk` that handles this property is [here](https://github.com/util-linux/util-linux/blob/28a43fc626b8dbe51270f8cf40a39c9ba714e70f/misc-utils/lsblk-properties.c#L336):

```c
struct lsblk_devprop *lsblk_device_get_properties(struct lsblk_device *dev)
{
	struct lsblk_devprop *p = NULL;

	DBG(DEV, ul_debugobj(dev, "%s: properties requested", dev->filename));
	if (lsblk->sysroot)
		return get_properties_by_file(dev);

	p = get_properties_by_udev(dev);
	if (!p)
		p = get_properties_by_blkid(dev);
	return p;
}
```

If you modify the arguments to `lsblk` to include `--sysroot /host` then the existing bind mount for `/host` seems to also be enough to correctly fetch the filesystem type because it looks like it uses an entirely different code path for fetching device properties.

**Comment by travisghansen** [comment]
Hmm. So begs the question why was it working in v1.4.2 if those files are used for both images? I did update the base debian image for v1.6.0 so maybe something changed in there…

**Comment by travisghansen** [comment]
Can someone test the `next` image to see if it errors out instead of reformatting the drive?

**Comment by Cellivar** [comment]
> Can someone test the `next` image to see if it errors out instead of reformatting the drive?

We'll have some bandwidth to experiment with it this weekend, yes.

**Comment by travisghansen** [comment]
This code is now live in `latest` and `v1.7.3` as well now.

**Comment by TripleE-0** [comment]
Sorry Travis, I didn't get to test until today. 
When I remove the mounts `/sys` and `/run/udev` in the node container, it consistently wipes the iscsi drive when using the "next" docker image and the "v1.7.3" image. The container runs successfully but does not have the data. 
Do you want any logs and if so which ones?

**Comment by travisghansen** [comment]
Hmm, that really doesn't make much sense. I'll do some further testing on my side before asking for further info.

**Comment by travisghansen** [comment]
@binaryf0x in your previous comment you mentioned the `property comes back as null`. Do you mean the `blockdevices` array is empty? or the array has 1 entry but the `fstype` property is empty?

I mean this from running `lsblk -a -b -J -O /dev/sdd`

**Comment by binaryf0x** [comment]
> @binaryf0x in your previous comment you mentioned the `property comes back as null`. Do you mean the `blockdevices` array is empty? or the array has 1 entry but the `fstype` property is empty?

The array has one entry and the `fstype` property is `null`.

**Comment by travisghansen** [comment]
Oh that’s really bad. @TripleE-0 is that what’s happening in your case? I don’t see how the code will format currently if there are not any entries in the array..

**Comment by TripleE-0** [comment]
Looks like you are correct @travisghansen, the fstype property is null. 
This is the result that I get when I run the lsblk -a -b -J -O /dev/sdc.
```
{
  "blockdevices": [
    {
      "name": "sdc",
      "kname": "sdc",
      "path": "/dev/sdc",
      "maj:min": "8:32",
      "fsavail": "1842257920",
      "fssize": "2040373248",
      "fstype": null,
      "fsused": "73965568",
      "fsuse%": "4%",
      "fsver": null,
      "mountpoint": "/host/var/lib/nomad/client/csi/node/org.democratic-csi.iscsi/per-alloc/592453f8-94c9-441d-4748-b91d2a385dc7/wikijs-db-data/rw-file-system-single-node-writer",
      "label": null,
      "uuid": null,
      "ptuuid": null,
      "pttype": null,
      "parttype": null,
      "parttypename": null,
      "partlabel": null,
      "partuuid": null,
      "partflags": null,
      "ra": 4096,
      "ro": false,
      "rm": false,
      "hotplug": false,
      "model": "iSCSI Disk      ",
      "serial": null,
      "size": 2147483648,
      "state": "running",
      "owner": "root",
      "group": "disk",
      "mode": "brw-rw----",
      "alignment": 0,
      "min-io": 16384,
      "opt-io": 16384,
      "phy-sec": 16384,
      "log-sec": 4096,
      "rota": true,
      "sched": "mq-deadline",
      "rq-size": 226,
      "type": "disk",
      "disc-aln": 0,
      "disc-gran": 16384,
      "disc-max": 268435456,
      "disc-zero": false,
      "wsame": 268435456,
      "wwn": null,
      "rand": true,
      "pkname": null,
      "hctl": "34:0:0:0",
      "tran": "iscsi",
      "subsystems": "block:scsi",
      "rev": "360 ",
      "vendor": "TrueNAS ",
      "zoned": "none",
      "dax": false
    }
  ]
}
```

**Comment by travisghansen** [comment]
Yeah ok. In the initial research I understood nothing was being returned at all for the device. I’m not sure there’s much that can be done on this one but I’ll do some more research.

In the container without the added mounts is there anything in `/run/udev` for you? Does the folder even exist?

**Comment by TripleE-0** [comment]
That folder does not exist. I see /run/lock and /run/utmp.

**Comment by travisghansen** [comment]
Ok. I suppose I can check for the existence of that folder and error out if it’s not there…or use a different tool/interface altogether.

**Comment by binaryf0x** [comment]
As mentioned previously it looks like the `--sysroot` option offers a workaround:

> If you modify the arguments to `lsblk` to include `--sysroot /host` then the existing bind mount for `/host` seems to also be enough to correctly fetch the filesystem type because it looks like it uses an entirely different code path for fetching device properties.

**Comment by travisghansen** [comment]
Yeah I remember that, but currently none of the core codebase depends on `/host` and I’d prefer to keep it that way. It is an option however and I won’t rule out detecting the presence of that directory etc. Even in that case I prefer to simply document the need for the udev mount vs the alternative.

Do you know if that simply changes where it looks for udev data or does it do something else altogether?

**Comment by binaryf0x** [comment]
It looks like it makes it does something else entirely.

**Comment by travisghansen** [comment]
I wonder if it’s possible to trigger that code path without the option? Or giving the value of `/`?

**Comment by TripleE-0** [comment]
I just tried the `--sysroot /host` option and it doesn't appear to supply the fstype information either. 


<details>
    <summary>lsblk sysroot results</summary>

```
root@nomad4:/home/csi/app# lsblk --sysroot /host -a -b -J -O /dev/sdc
{
  "blockdevices": [
    {
      "name": "sdc",
      "kname": "sdc",
      "path": "/dev/sdc",
      "maj:min": "8:32",
      "fsavail": "1842257920",
      "fssize": "2040373248",
      "fstype": null,
      "fsused": "73965568",
      "fsuse%": "4%",
      "fsver": null,
      "mountpoint": "/host/var/lib/nomad/client/csi/node/org.democratic-csi.iscsi/per-alloc/592453f8-94c9-441d-4748-b91d2a385dc7/wikijs-db-data/rw-file-system-single-node-writer",
      "label": null,
      "uuid": null,
      "ptuuid": null,
      "pttype": null,
      "parttype": null,
      "parttypename": null,
      "partlabel": null,
      "partuuid": null,
      "partflags": null,
      "ra": 4096,
      "ro": false,
      "rm": false,
      "hotplug": false,
      "model": "iSCSI Disk      ",
      "serial": null,
      "size": 2147483648,
      "state": "running",
      "owner": "root",
      "group": "disk",
      "mode": "brw-rw----",
      "alignment": 0,
      "min-io": 16384,
      "opt-io": 16384,
      "phy-sec": 16384,
      "log-sec": 4096,
      "rota": true,
      "sched": "mq-deadline",
      "rq-size": 226,
      "type": "disk",
      "disc-aln": 0,
      "disc-gran": 16384,
      "disc-max": 268435456,
      "disc-zero": false,
      "wsame": 268435456,
      "wwn": null,
      "rand": true,
      "pkname": null,
      "hctl": "34:0:0:0",
      "tran": "iscsi",
      "subsystems": "block:scsi",
      "rev": "360 ",
      "vendor": "TrueNAS ",
      "zoned": "none",
      "dax": false
    }
  ]
}
```
</details>

**Comment by binaryf0x** [comment]
It feels like using `/` would be incorrect though. We're trying to read the state of the host, not the container.

**Comment by travisghansen** [comment]
> It feels like using `/` would be incorrect though. We're trying to read the state of the host, not the container.

Sure, but the container is privileged and does implicitly have access to substantial amounts of host data already.

**Comment by travisghansen** [comment]
> I just tried the `--sysroot /host` option and it doesn't appear to supply the fstype information either.

Can you try `—sysroot /`?

**Comment by TripleE-0** [comment]
> > I just tried the `--sysroot /host` option and it doesn't appear to supply the fstype information either.
> 
> Can you try `—sysroot /`?

`lsblk --sysroot / -a -J -O /dev/sdc` gives the same fstype: null value.

**Comment by travisghansen** [comment]
Can you try command (without the relevant mounts in the container)?

```
blkid --probe <device>
```

**Comment by TripleE-0** [comment]
That seems to work, it returns the filesystem type at least.

```
root@nomad2:/home/csi/app# lsblk -a -J -O /dev/sdd
{
   "blockdevices": [
      {"name":"sdd", "kname":"sdd", "path":"/dev/sdd", "maj:min":"8:48", "fsavail":"906.1M", "fssize":"973.4M", "fstype":null, "fsused":"116K", "fsuse%":"0%", "fsver":null, "mountpoint":"/host/var/lib/nomad/client/csi/node/org.democratic-csi.iscsi/per-alloc/1be1cf8f-a181-dcb1-00cd-eec4a05cb3ed/traefik/rw-file-system-single-node-writer", "label":null, "uuid":null, "ptuuid":null, "pttype":null, "parttype":null, "parttypename":null, "partlabel":null, "partuuid":null, "partflags":null, "ra":4096, "ro":false, "rm":false, "hotplug":false, "model":"iSCSI Disk      ", "serial":null, "size":"1G", "state":"running", "owner":"root", "group":"disk", "mode":"brw-rw----", "alignment":0, "min-io":16384, "opt-io":16384, "phy-sec":16384, "log-sec":4096, "rota":true, "sched":"mq-deadline", "rq-size":226, "type":"disk", "disc-aln":0, "disc-gran":"16K", "disc-max":"256M", "disc-zero":false, "wsame":"256M", "wwn":null, "rand":true, "pkname":null, "hctl":"35:0:0:0", "tran":"iscsi", "subsystems":"block:scsi", "rev":"360 ", "vendor":"TrueNAS ", "zoned":"none", "dax":false}
   ]
}
root@nomad2:/home/csi/app# blkid --probe /dev/sdd
/dev/sdd: UUID="7bd76abe-3619-4d5f-b190-03b73f46335c" VERSION="1.0" BLOCK_SIZE="4096" TYPE="ext4" USAGE="filesystem"

```

**Comment by travisghansen** [comment]
I have added an ability to tweak this behavior, but am unlikely to alter the default for the time being: https://github.com/democratic-csi/democratic-csi/commit/516701bd29a41f4b6ffe0b90d26e286c5196706c

If you try the new `next` images and set the `FILESYSTEM_TYPE_DETECTION_STRATEGY=blkid` env var in the node deployment it should use `blkid` for the check instead of `lsblk`. I don't really love the `blkid` feature because if the block device has no partition table and isn't formatted the command exits with a `2` exit code. I have no idea if other scenarios could lead to that exit code which is why I'm a bit hesitant to go that route by default.

At this point I think we're best to simply properly document the need for the udev mount and make sure it is added to all the examples etc.

**Comment by travisghansen** [comment]
Can someone create a PR with updated examples for `Nomad` to make sure we've got `udev` in there?

**Comment by dcarbone** [comment]
Unfortunately I am continuing to run into this, even after the updates suggested in #297.  It works fine a few times, but it seems like if a node where the volume is mounted restarts without a clean dismount, something gets "confused" and my mounts are invariably wiped :\

**Comment by travisghansen** [comment]
Are you mounting udev? Setting the env var? Both?

**Comment by dcarbone** [comment]
> Are you mounting udev? Setting the env var? Both?

Both :/ 

another wrinkle may be my infra consists of mixed archs. i run ubuntu server 22.04 x64 on all, but its a mix of pi4's and ancient (read: cheap) x86 machines. i haven't had enough time to debug whether that is relevant.

**Comment by travisghansen** [comment]
Remove the ENV var and just mount udev and let me know how it goes...

**Comment by dcarbone** [comment]
ah, that seems to do the trick!  several stops and restarts later, everything is mounting and data is persisting.

awesome :)

**Comment by travisghansen** [comment]
Yeah when I introduced the alternative option to detect filesystem I was afraid something crazy like that would show up eventually :( if you think you can figure out the scenario of when it’s not detecting properly so I can fix it or remove the feature altogether (that’s clearly a very very bad outcome) I would greatly appreciate the help/info.
